onion_message/onion_endpoint.go

```diff
@@ -65,21 +148,126 @@ func (o *OnionEndpoint) SendMessage(ctx context.Context,
 	}
 
 	peer := msg.PeerPub.SerializeCompressed()
-	log.Debugf("OnionEndpoint received OnionMessage from peer %s: "+
-		"BlindingPoint=%v, OnionPacket[:10]=%x...", peer,
+	log.Debugf("OnionEndpoint received OnionMessage from peer %x: "+
+		"BlindingPoint=%v, OnionPacket[:10]=%10x...", peer,
 		onionMsg.BlindingPoint, onionMsg.OnionBlob)
 
+	payload, err := o.handleOnionMessage(*onionMsg)
+	if err != nil {
+		log.Errorf("Failed to handle onion message: %v", err)
+	}
+
 	var peerArr [33]byte
 	copy(peerArr[:], peer)
-	err := o.onionMessageServer.SendUpdate(&OnionMessageUpdate{
+
+	update := &OnionMessageUpdate{
 		Peer:          peerArr,
 		BlindingPoint: onionMsg.BlindingPoint.SerializeCompressed(),
 		OnionBlob:     onionMsg.OnionBlob,
-	})
+	}
+
+	// If we have a payload (no error), add its contents to our update.
+	if payload != nil {
+		update.CustomRecords = payload.CustomRecords()
+		update.ReplyPath = payload.ReplyPath()
+		update.EncryptedRecipientData = payload.EncryptedData()
+	}
+
+	// Send the update to any subscribers.
+	if sendErr := o.onionMessageServer.SendUpdate(update); sendErr != nil {
+		log.Errorf("Failed to send onion message update: %v", sendErr)
+		return false
+	}
+
+	// If we failed to handle the onion message, we return false.
 	if err != nil {
-		log.Errorf("Failed to send onion message update: %v", err)
 		return false
 	}
 
 	return true
 }
+
+// handleOnionMessage decodes and processes an onion message.
+func (o *OnionEndpoint) handleOnionMessage(msg lnwire.OnionMessage) (
+	*hop.Payload, error) {
+	reqs := []hop.DecodeHopIteratorRequest{
+		{
+			OnionReader:    bytes.NewReader(msg.OnionBlob),
+			RHash:          nil,
+			IncomingCltv:   0,
+			IncomingAmount: 0,
+			BlindingPoint: tlv.SomeRecordT(
+				tlv.NewPrimitiveRecord[lnwire.BlindingPointTlvType](
+					msg.BlindingPoint,
+				),
+			),
+			IsOnionMessage: true,
+		},
+	}
+
+	resps, err := o.onionProcessor.DecodeHopIterators(
```

Comment:
IIUC, this'll also actually cause us to hit disk for replay protection. Thinking about it a bit more: is any replay protection even defined for onion messaging?

Reply:
yeah no need for reply protection.